File: calc/nqueens.go
package main

import (
	"fmt"
	"testing"
	"time"
)

func BenchmarkNQueens(b *testing.B) {
	n := 12

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = nQueens(n)
	}
}

func nQueens(n int) int {
	solutions := 0
	board := make([]int, n)
	placeQueen(board, 0, &solutions)
	return solutions
}

func placeQueen(board []int, currentRow int, solutions *int) {
	n := len(board)
	if currentRow == n {
		*solutions++
	} else {
		for col := 0; col < n; col++ {
			if canPlaceQueen(board, currentRow, col) {
				board[currentRow] = col
				placeQueen(board, currentRow+1, solutions)
			}
		}
	}
}

func canPlaceQueen(board []int, row, col int) bool {
	for i := 0; i < row; i++ {
		if board[i] == col || abs(row-i) == abs(col-board[i]) {
			return false
		}
	}
	return true
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	result := testing.Benchmark(BenchmarkNQueens)
	d := time.Duration(result.NsPerOp())
	fmt.Printf("%s\n", result)
	fmt.Printf("%s/op\n", d)
}

File: rust-calc/benches/nqueens.rs
use criterion::{black_box, criterion_group, criterion_main, Criterion};

fn n_queens(n: usize) -> usize {
    let mut solutions = 0;
    let mut board = vec![0; n];
    place_queen(&mut board, 0usize, &mut solutions);
    solutions
}

fn place_queen(board: &mut Vec<usize>, current_row: usize, solutions: &mut usize) {
    let n = board.len();
    if current_row == n {
        *solutions += 1;
    } else {
        for col in 0..n {
            if can_place_queen(&board, current_row, col) {
                board[current_row] = col;
                // unsafe { *board.get_unchecked_mut(current_row) = col }
                place_queen(board, current_row + 1, solutions);
            }
        }
    }
}

fn can_place_queen(board: &[usize], row: usize, col: usize) -> bool {
    for i in 0..row {
        if board[i] == col || (row - i) == (col as isize - board[i] as isize).abs() as usize {
            return false;
        }
    }
    true
}

fn bench_n_queens(c: &mut Criterion) {
    c.bench_function("n_queens", |b| {
        b.iter(|| n_queens(12));
    });

    // c.bench_function("n_queens_blackbox", |b| {
    //     b.iter(|| n_queens(black_box(12)));
    // });
}

criterion_group!(benches, bench_n_queens);
criterion_main!(benches);
